### DISCUSSION TITLE
The Roadmap - I need your help!

### DISCUSSION THREAD

**Discussion by darrenburns** [darrenburns/posting#7] [Announcements]
Hi everyone,

I'm thinking about Posting's roadmap, and would love your help!

If you have any ideas for features, UX improvements, or anything else, please open up a discussion!

I'm open to hear your ideas about anything - they don't need to be grand plans, even simple suggestions are much appreciated. I want to make Posting great for all developers, and I can't do it alone!

## Planned features

These are just some of the things I'm considering. The priority is currently undefined. If you have any opinions on these I'd love to chat!

- **Help system**: Posting has a bunch of keybindings, and most of them aren't displayed in the footer. I'd like to investigate a dynamic/contextual help system which can display relevant help/keybinding information based on where you currently are in the interface. ✅
- **Validation**: You can currently do things you shouldn't be able to. Empty request names, etc. 🟠
- **Request Auth tab**: Attach/edit auth to outgoing requests ✅.
- **Request options persisted**: Right now, request options are a runtime thing - but they should be persisted on disk inside the posting yaml file (as you might think from the current UI) ✅.
- **Cookie jar**: More control over cookies in a session.
- **Resizing sections**: I've implemented the ability to show and hide sections, including the request and response sections and the collection browser sidebar ✅.
- **Other body types**: Sending form data, files, etc 🟠.
- **Configuration**: simple configuration file loaded on startup. Will allow setting initial values, and things like setting the default theme etc. I'd like to also support per-collection configuration files. So you can e.g. set some config that overrides the global config for a single collection 🟠.
- **Proxy support**: Add support for proxies ✅.
- **Environments**: define variables in a `yaml` file and load them into Posting. You can then refer to these variables in the URL, request body, header values, etc. ✅
- **Importing**: converting from OpenAPI, Postman collections, etc. into `posting.yaml` files. 🟠
- **Duplicating requests**: quickly create a copy of a request. ✅
- **Testing**: run assertions on receiving responses.

✅ = implemented
🟠 = partially implemented

Cheers!

Darren

**Comment by cemrehancavdar** [comment]
Pre and post request scripting ability is the first thing i check on Postman alternatives. It simplifies things so much and could be perfect to have. I would really love pythonic one.

**Reply by darrenburns** [reply]
What kinds of things do you use this for? 🙂

**Reply by cemrehancavdar** [reply]
We have implementation with various APIs for our work. One of them takes body as JSON AES encrypted.

Here is the pre-request script for encryption:
``` js
const secretKey = 'some-secret key';
const encryptionConfig = {
  mode: CryptoJS.mode.ECB,
  padding: CryptoJS.pad.Pkcs7,
  blockSize: 16,
};

const encryptedBody = CryptoJS.AES.encrypt(
  pm.request.body.raw,
  CryptoJS.enc.Utf8.parse(secretKey),
  encryptionConfig
).toString();
pm.request.body.raw = {value: `${encryptedBody}`}
```

I also use post request scripts to set/change environment variables. It simplifies testing several API's. For e.g: login (get and set auth token), create order (get and set order_id), change order status (use order_id env param to do this) etc.

**Reply by darrenburns** [reply]
This is great information, thank you. I'll almost certainly add something like this - but will need some time (and suggestions!) to think of a nice approach.

I'm thinking you can attach a script (somehow) that looks something like this:

```python
import httpx

def before(request: httpx.Request) -> httpx.Request:
    """Modify the request, returning the new request."""

def after(response: httpx.Response) -> None:
    """Do things with the response"""
```

I have no clue how virtual environments and extra dependencies would work, or how the scripts would be attached to requests on the file system though.

I was also thinking of allowing test scripts, but thinking they should be a separate entity rather than writing assertions inside a post-request script though.

Mostly just dumping very raw/initial thoughts here :)

**Reply by tusharsadhwani** [reply]
I think allowing external dependencies is a bit overkill at least at this stage, but `yen` does solve having to create venvs for packages with a certain python version etc. if you need to use it.

**Reply by darrenburns** [reply]
You're probably right!

**Comment by tusharsadhwani** [comment]
Not just the sidebar, can the output pane for example also be resizable? Maybe even collapsible and expandable to full height?

**Reply by darrenburns** [reply]
Yes, I was thinking of having a keybind which grows the size of the currently focused section e.g. the request/response section.

**Reply by tusharsadhwani** [reply]
ctrl+F for full screen perhaps? or ctrl+M for maximize (is what I use on vscode since it's not already used by some other shortcut)

**Reply by treyhunner** [reply]
This is my second-most desired feature. Maximizing or full screening that output pane to take up the full width and height of the terminal would be amazing! 👍

**Reply by treyhunner** [reply]
Just a note that this is how posting looks in vertical and horizontal mode for me with the font size that I use while teaching:

![image](https://github.com/darrenburns/posting/assets/285352/093a357c-06fd-4418-8a40-3ff96194ff76)

![image](https://github.com/darrenburns/posting/assets/285352/76e67a99-7604-45a3-8f79-06e07bd0aedc)

It's usable even in this form, but I happened to only need to show about 3 lines of text on the screen at a time. A way to minimize/hide the collection pane, request pane, title bar, etc. would help a lot.

I'm excited for how quickly this project is becoming a very usable application for me. 👏

**Reply by darrenburns** [reply]
Cheers! I just released 1.0.0b27, where you can press `ctrl+m` with the request/response section active and it will maximise that section. Press that key combo again to set the request and response back to equal widths.

**Reply by treyhunner** [reply]
I just upgraded to 1.0.0b27 and Ctrl-M didn't seem to do anything when the request or response sections were selected.

Here's what my screen looked like:

![image](https://github.com/darrenburns/posting/assets/285352/dd739f25-69be-4d26-9cf1-8fc716b47021)

I'm not sure if there's something I need to do special besides clicking on those sections.

**Reply by darrenburns** [reply]
Could you run `textual keys` and press `ctrl+m` and let me now what it prints? I wonder if your emulator is producing something else.

**Reply by treyhunner** [reply]
> Could you run `textual keys` and press `ctrl+m` and let me now what it prints? I wonder if your emulator is producing something else.

Doing that seems to press the big yellow "Clear" button.

When I held it down, the screen shows a few rows and flashes (clearing itself sometimes), but I managed to capture this:

```
Key(key='enter', character='\r', name='enter', is_printable=False, aliases=['enter', 'ctrl+m'])
Key(key='enter', character='\r', name='enter', is_printable=False, aliases=['enter', 'ctrl+m'])
Key(key='enter', character='\r', name='enter', is_printable=False, aliases=['enter', 'ctrl+m'])
Key(key='enter', character='\r', name='enter', is_printable=False, aliases=['enter', 'ctrl+m'])
Key(key='enter', character='\r', name='enter', is_printable=False, aliases=['enter', 'ctrl+m'])
```

**Edit**: `Enter` had the same issue and I realized that when I launch `textual keys` that the Clear button is selected. When I manually click in the middle of my terminal and then press either `Enter` or `Ctrl+M` I see the same thing:

```
Key(key='enter', character='\r', name='enter', is_printable=False, aliases=['enter', 'ctrl+m'])
```

**Reply by darrenburns** [reply]
I'm guessing you're using an older terminal emulator? I think `ctrl+m` is only working on emulators which support the Kitty protocol 😢 

I'll try to find another keybinding for this that doesn't clash. I'll also put it into the command palette.

**Reply by treyhunner** [reply]
I'm using tmux in Gnome Terminal.
I don't tmux is related though because I just opened up XTerm without using tmux and it has the same issue.

**Edit**: just confirmed that Gnome Terminal without tmux running causes the same issue.

**Reply by darrenburns** [reply]
I've improved the command palette with some new view options (and also made it a bit smarter about not showing options that are irrelevant based on context).

There are new "view" options in there, for expanding the request and response areas.

This is in the latest version. I'll have a think about the keybindings in the meantime :)

**Reply by darrenburns** [reply]
Added a binding and command palette option for the sidebar now too, so sections can now be made pretty much full screen!

https://github.com/darrenburns/posting/assets/5740731/36b6ed14-21f4-4e19-ad83-0f63bbf80ee1

Also, there's a config option you can add to your `config.yaml` for hiding the app header if you're really struggling for space (note that I might update the config structure a little bit before the final 1.0 release, but this will work for now):

```yaml
heading:
  visible: false
```

**Comment by tracure1337** [comment]
Love the Roadmap. Love the project. Love u!
Maybe the json editor can also mirror a visualization a la jsoncrack. Otherweise the bag of ideas is pretty and nicely filled.
I'll wait a few more versions before I might try to add to it.

**Reply by tracure1337** [reply]
I also think a lot of cues can be taken from tools like Caido ,BurpSuite, Insomnia & Bruno.

**Reply by darrenburns** [reply]
Thank you!

Is there anything in particular from these tools that you enjoy?

**Comment by treyhunner** [comment]
> * **Configuration**: simple configuration file loaded on startup. Will allow setting initial values, and things like setting the default theme etc. I'd like to also support per-collection configuration files. So you can e.g. set some config that overrides the global config for a single collection.

This is the feature I'm really waiting for.

Even if there was a `~/.posting-config-temp-will-be-renamed.toml` file that will certainly be renamed and with a format that will be modified in future releases, I'd love *some sort* of configuration file.

The main thing I'd want to put in that file right now is my preferred color scheme.
I've been using this while teaching over the last couple weeks and it's a bit cumbersome to switch the color scheme each time I load posting.

**Reply by darrenburns** [reply]
Config will come soon. It's probably gonna be the next thing I work on after I get back from my holiday 🙂
Possibly this week.

**Reply by darrenburns** [reply]
Configuration is in the latest release!

https://github.com/darrenburns/posting/blob/main/README.md#configuration

**Comment by tusharsadhwani** [comment]
hey just to note, posting doesn't install on macos on python 3.12. there's no issues page to report this on, so just making sure you are aware of this.

it's because of brotlipy not building on the platform

**Reply by tusharsadhwani** [reply]
it does build if you set the environment variable `AR=/usr/bin/ar` so building wheels with that setup on ci should fix that.

**Reply by darrenburns** [reply]
The latest version should work fine as far as I'm aware?

**Reply by darrenburns** [reply]
(brotlipy is no longer a dependency)

**Reply by darrenburns** [reply]
(Also I've opened up the Issues page now)

**Reply by tusharsadhwani** [reply]
ah, great.

**Comment by davidfokkema** [comment]
Not sure if you want feature requests here or in the issue tracker, but I'd _love_ the option to pretty print the json response. I got a 10k+ character response and it's quite difficult to visually parse the different elements.

**Reply by darrenburns** [reply]
I prefer feature requests as "discussions" with the "idea" label. I should probably try to communicate this 😄 

Anyway, pretty-printing JSON responses has been added by @tusharsadhwani and is now enabled by default in the latest release.

**Reply by davidfokkema** [reply]
Yes, this works great, thanks @tusharsadhwani! And @darrenburns, I belatedly _do_ see "please open up a discussion!" in your post at the top. Will do, next time!

**Comment by felipeospina21** [comment]
I was wondering if there is a plan to support graphql queries with syntax highlight and autocomplete? By the way thanks for this amazing project it is looking amazing so far.

**Reply by darrenburns** [reply]
No plans for that at the moment, but if I get enough interest I might consider it 😄

**Reply by felipeospina21** [reply]
It would be great to have it at some point so we can move away from insomnia. 😎

**Reply by adriangalilea** [reply]
Would be great to have a dedicated issue/discussion to subscribe to for the graphql feature specifically

**Reply by darrenburns** [reply]
Feel free to create an "idea" discussion 🙂

**Reply by felipeospina21** [reply]
I created it https://github.com/darrenburns/posting/discussions/89

**Reply by adriangalilea** [reply]
damn me too lol https://github.com/darrenburns/posting/discussions/90 (closed it 😄 )

**Reply by darrenburns** [reply]
Thanks! Let's go with the first one created (by mere seconds) #89! 😂

**Comment by ArkadyBuryakov** [comment]
Websockets support would be super cool:
1. Open WS connections, send messages with variables into channel, see the channel log.
2. Save outgoing messages to library for frequent usage.
